From 075c16e0a65a70249108d73a039d6cc6eda80cf0 Mon Sep 17 00:00:00 2001
From: Dmitry Muhomor <muhomor.dmitry@gmail.com>
Date: Wed, 29 May 2024 12:08:48 +0000
Subject: [PATCH] add duress password management UI

Ported from GrapheneOS Settings 2025061900 (39f2f7eedb). AXP fork: strings moved to calyx_strings.xml (AXP.OS instead of GrapheneOS), entry under Settings > Security instead of Safety Center.

Change-Id: I19def9da5ddd85ec0e1aa000dfa461e24dbfa542
---
 AndroidManifest.xml                           |  20 ++
 res/layout/duress_password_setup.xml          |  67 ++++++
 res/values/calyx_strings.xml                  |  27 +++
 res/xml/security_dashboard_settings.xml       |   9 +
 .../security/DuressPasswordActivity.java      |  71 +++++++
 .../security/DuressPasswordMainActivity.java  | 182 ++++++++++++++++
 .../security/DuressPasswordSetupActivity.java | 196 ++++++++++++++++++
 7 files changed, 572 insertions(+)
 create mode 100644 res/layout/duress_password_setup.xml
 create mode 100644 src/com/android/settings/security/DuressPasswordActivity.java
 create mode 100644 src/com/android/settings/security/DuressPasswordMainActivity.java
 create mode 100644 src/com/android/settings/security/DuressPasswordSetupActivity.java

diff --git a/AndroidManifest.xml b/AndroidManifest.xml
index 0155131..0c3aa10 100644
--- a/AndroidManifest.xml
+++ b/AndroidManifest.xml
@@ -3016,6 +3016,26 @@
             <meta-data android:name="com.android.settings.icon_tintable" android:value="true" />
         </activity>
 
+        <activity android:name=".security.DuressPasswordMainActivity"
+            android:excludeFromRecents="true"
+            android:theme="@style/GlifTheme.Light"
+            android:permission="android.permission.MANAGE_SAFETY_CENTER"
+            android:exported="true">
+
+            <intent-filter>
+                <action android:name="com.android.settings.DURESS_PASSWORD_SETTINGS" />
+                <category android:name="android.intent.category.DEFAULT" />
+            </intent-filter>
+
+            <meta-data android:name="com.android.settings.HIGHLIGHT_MENU_KEY"
+                       android:value="@string/menu_key_safety_center"/>
+        </activity>
+
+        <activity android:name=".security.DuressPasswordSetupActivity"
+            android:excludeFromRecents="true"
+            android:theme="@style/GlifTheme.Light"
+            android:exported="false" />
+
         <activity android:name=".biometrics.fingerprint.FingerprintEnrollSuggestionActivity"
             android:exported="true"
             android:icon="@drawable/ic_suggestion_fingerprint">
diff --git a/res/layout/duress_password_setup.xml b/res/layout/duress_password_setup.xml
new file mode 100644
index 0000000..a708b3f
--- /dev/null
+++ b/res/layout/duress_password_setup.xml
@@ -0,0 +1,67 @@
+<?xml version="1.0" encoding="utf-8"?>
+
+<com.google.android.setupdesign.GlifLayout
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:id="@+id/glif_layout"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:icon="@drawable/ic_lock"
+>
+    <LinearLayout
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:orientation="vertical"
+        style="@style/SudContentFrame">
+
+        <TextView
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:textAppearance="@style/TextAppearance.Material3.TitleMedium"
+            android:text="@string/unlock_set_unlock_pin_title" />
+
+        <EditText
+            android:id="@+id/pin_input"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="16dp"
+            android:hint="@string/duress_pwd_enter_pin"
+            android:imeOptions="actionNext"
+            android:inputType="numberPassword" />
+
+        <EditText
+            android:id="@+id/pin_input_confirmation"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="16dp"
+            android:hint="@string/duress_pwd_confirm_pin"
+            android:imeOptions="actionNext"
+            android:inputType="numberPassword" />
+
+        <TextView
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="32dp"
+            android:textAppearance="@style/TextAppearance.Material3.TitleMedium"
+            android:text="@string/unlock_set_unlock_password_title" />
+
+        <EditText
+            android:id="@+id/password_input"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="16dp"
+            android:hint="@string/duress_pwd_enter_password"
+            android:imeOptions="actionNext"
+            android:inputType="textPassword" />
+
+        <EditText
+            android:id="@+id/password_input_confirmation"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="16dp"
+            android:hint="@string/duress_pwd_confirm_password"
+            android:imeOptions="actionDone"
+            android:inputType="textPassword" />
+
+    </LinearLayout>
+</com.google.android.setupdesign.GlifLayout>
diff --git a/res/values/calyx_strings.xml b/res/values/calyx_strings.xml
index bb689a6..8b49cc3 100644
--- a/res/values/calyx_strings.xml
+++ b/res/values/calyx_strings.xml
@@ -33,4 +33,31 @@
     <!-- AXP fork: deny new USB devices while locked -->
     <string name="deny_new_usb_title">Block new USB devices when locked</string>
     <string name="deny_new_usb_summary">USB devices connected while the phone is locked are ignored until you unlock and reconnect them. Charging is not affected.</string>
+    <!-- AXP fork: duress PIN/password (GrapheneOS) -->
+    <string name="duress_pwd_pref_title">Duress password</string>
+    <string name="duress_pwd_description">"Duress password can be entered instead of the user password in any place that asks for it, such as the lock screen of any user and any user or work profile password prompt.
+
+Submitting the duress password will immediately make the storage contents permanently inaccessible, delete all eSIMs and then power off the device. AXP.OS will remain installed.
+
+Duress PIN works the same way duress password does."</string>
+    <string name="duress_pwd_enter_pin">Enter duress PIN</string>
+    <string name="duress_pwd_confirm_pin">Confirm duress PIN</string>
+    <string name="duress_pwd_enter_password">Enter duress password</string>
+    <string name="duress_pwd_confirm_password">Confirm duress password</string>
+    <string name="duress_pwd_action_add">Add duress PIN and password</string>
+    <string name="duress_pwd_add_button">Add</string>
+    <string name="duress_pwd_toast_added">Added duress PIN and password</string>
+    <string name="duress_pwd_toast_updated">Updated duress PIN and password</string>
+    <string name="duress_pwd_update_button">Update</string>
+    <string name="duress_pwd_cancel_button">Cancel</string>
+    <string name="duress_pwd_proceed_button">Proceed</string>
+    <string name="duress_pwd_delete_button">Delete</string>
+    <string name="duress_pwd_action_update">Update duress PIN and password</string>
+    <string name="duress_pwd_action_delete">Delete duress PIN and password</string>
+    <string name="duress_pwd_action_delete_confirmation">Delete duress PIN and password?</string>
+    <string name="duress_pwd_save_warning_title">WARNING</string>
+    <string name="duress_pwd_save_warning_text">"Submitting the duress PIN/password instead of the user PIN/password <b>will immediately make the storage contents permanently inaccessible</b> and will delete all eSIMs."</string>
+    <string name="duress_pwd_save_error">Duress PIN and password were not saved.\n\nDetails:\n%1$s</string>
+    <string name="duress_pwd_delete_error">Duress PIN and password were not deleted.\n\nDetails:\n%1$s</string>
+    <string name="duress_pwd_error_dialog_dismiss">Dismiss</string>
 </resources>
diff --git a/res/xml/security_dashboard_settings.xml b/res/xml/security_dashboard_settings.xml
index 03763b7..57c8fd5 100644
--- a/res/xml/security_dashboard_settings.xml
+++ b/res/xml/security_dashboard_settings.xml
@@ -76,6 +76,15 @@
             android:summary="@string/deny_new_usb_summary"
             settings:controller="com.android.settings.security.DenyNewUsbPreferenceController" />
 
+        <!-- AXP fork: duress PIN/password (GrapheneOS keeps it in Safety Center) -->
+        <Preference
+            android:key="duress_password"
+            android:title="@string/duress_pwd_pref_title">
+            <intent
+                android:action="com.android.settings.DURESS_PASSWORD_SETTINGS"
+                android:targetPackage="com.android.settings" />
+        </Preference>
+
         <SwitchPreference
             android:key="exec_spawn"
             android:title="@string/exec_spawn_title"
diff --git a/src/com/android/settings/security/DuressPasswordActivity.java b/src/com/android/settings/security/DuressPasswordActivity.java
new file mode 100644
index 0000000..54a6cb8
--- /dev/null
+++ b/src/com/android/settings/security/DuressPasswordActivity.java
@@ -0,0 +1,71 @@
+package com.android.settings.security;
+
+import android.app.Activity;
+import android.os.Bundle;
+import android.view.WindowManager;
+
+import androidx.annotation.Nullable;
+
+import com.android.internal.widget.LockPatternUtils;
+import com.android.settings.SetupWizardUtils;
+import com.android.settings.overlay.FeatureFactory;
+import com.google.android.setupdesign.GlifLayout;
+import com.google.android.setupdesign.template.DescriptionMixin;
+import com.google.android.setupdesign.util.ThemeHelper;
+
+public class DuressPasswordActivity extends Activity {
+
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        getWindow().addFlags(WindowManager.LayoutParams.FLAG_SECURE);
+
+        setTheme(SetupWizardUtils.getTheme(this, getIntent()));
+        ThemeHelper.trySetDynamicColor(this);
+
+        setResult(RESULT_OK);
+    }
+
+    protected boolean allowNextOnStop;
+
+    @Override
+    protected void onStop() {
+        super.onStop();
+        if (!isChangingConfigurations() && hasUserCredential()) {
+            if (allowNextOnStop) {
+                allowNextOnStop = false;
+            } else {
+                // require user credential to be re-entered after activity is backgrounded
+                setResult(RESULT_CANCELED);
+                finish();
+            }
+        }
+    }
+
+    protected boolean hasUserCredential() {
+        return true;
+    }
+
+    @Override
+    protected void onDestroy() {
+        super.onDestroy();
+        if (!isChangingConfigurations()) {
+            /** @see com.android.settings.password.ConfirmDeviceCredentialBaseActivity#onDestroy */
+            getMainThreadHandler().postDelayed(() -> {
+                System.gc();
+                System.runFinalization();
+                System.gc();
+            }, 5000);
+        }
+    }
+
+    static void adjustDescriptionStyle(GlifLayout l) {
+        l.getMixin(DescriptionMixin.class).getTextView().setTextSize(16f);
+    }
+
+    LockPatternUtils getLockPatternUtils() {
+        return FeatureFactory.getFeatureFactory()
+                .getSecurityFeatureProvider()
+                .getLockPatternUtils(this);
+    }
+}
diff --git a/src/com/android/settings/security/DuressPasswordMainActivity.java b/src/com/android/settings/security/DuressPasswordMainActivity.java
new file mode 100644
index 0000000..e99f38e
--- /dev/null
+++ b/src/com/android/settings/security/DuressPasswordMainActivity.java
@@ -0,0 +1,182 @@
+package com.android.settings.security;
+
+import android.app.AlertDialog;
+import android.content.Intent;
+import android.os.Bundle;
+import android.util.Log;
+
+import androidx.annotation.DrawableRes;
+import androidx.annotation.Nullable;
+import androidx.annotation.StringRes;
+
+import com.android.internal.widget.LockPatternUtils;
+import com.android.internal.widget.LockscreenCredential;
+import com.android.settings.R;
+import com.android.settings.password.ChooseLockSettingsHelper;
+import com.google.android.setupdesign.GlifRecyclerLayout;
+import com.google.android.setupdesign.items.IItem;
+import com.google.android.setupdesign.items.Item;
+import com.google.android.setupdesign.items.ItemGroup;
+import com.google.android.setupdesign.items.RecyclerItemAdapter;
+
+import java.util.Objects;
+
+import static java.util.Objects.requireNonNull;
+
+public class DuressPasswordMainActivity extends DuressPasswordActivity implements RecyclerItemAdapter.OnItemSelectedListener {
+    private static final String TAG = DuressPasswordMainActivity.class.getSimpleName();
+    private static final String KEY_USER_CREDENTIAL = "user_credential";
+    private static final String KEY_HAS_ASKED_FOR_USER_CREDENTIALS = "asked_for_user_credentials";
+
+    private GlifRecyclerLayout layout;
+    private LockscreenCredential userCredential;
+    private boolean askedForUserCredentials;
+
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        var layout = new GlifRecyclerLayout(this);
+        this.layout = layout;
+        layout.setIcon(getDrawable(R.drawable.ic_lock));
+        layout.setHeaderText(R.string.duress_pwd_pref_title);
+        adjustDescriptionStyle(layout);
+        layout.setDescriptionText(R.string.duress_pwd_description);
+
+        setContentView(layout);
+
+        if (savedInstanceState != null) {
+            userCredential = savedInstanceState.getParcelable(KEY_USER_CREDENTIAL, LockscreenCredential.class);
+            askedForUserCredentials = requireNonNull(savedInstanceState.getBoolean2(KEY_HAS_ASKED_FOR_USER_CREDENTIALS));
+        }
+    }
+
+    @Override
+    protected boolean hasUserCredential() {
+        return userCredential != null && !userCredential.isNone();
+    }
+
+    @Override
+    protected void onSaveInstanceState(Bundle outState) {
+        super.onSaveInstanceState(outState);
+        outState.putParcelable(KEY_USER_CREDENTIAL, userCredential);
+        outState.putBoolean(KEY_HAS_ASKED_FOR_USER_CREDENTIALS, askedForUserCredentials);
+    }
+
+    @Override
+    protected void onResume() {
+        super.onResume();
+
+        if (userCredential == null) {
+            if (!getLockPatternUtils().isSecure(getUserId())) {
+                userCredential = LockscreenCredential.createNone();
+            } else if (!askedForUserCredentials) {
+                var b = new ChooseLockSettingsHelper.Builder(this);
+                b.setRequestCode(REQ_CODE_OBTAIN_USER_CREDENTIALS);
+                b.setReturnCredentials(true);
+                b.setForegroundOnly(true);
+                b.show();
+                askedForUserCredentials = true;
+            }
+        }
+
+        updateActionList();
+    }
+
+    private void updateActionList() {
+        if (userCredential == null) {
+            Log.d(TAG, "no userCredential, skipping updateActionList");
+            return;
+        }
+
+        var g = new ItemGroup();
+
+        if (getLockPatternUtils().hasDuressCredentials(userCredential)) {
+            g.addChild(createItem(R.string.duress_pwd_action_update, R.drawable.ic_edit));
+            g.addChild(createItem(R.string.duress_pwd_action_delete, R.drawable.ic_delete));
+        } else {
+            g.addChild(createItem(R.string.duress_pwd_action_add, R.drawable.ic_add_24dp));
+        }
+
+        var adapter = new RecyclerItemAdapter(g);
+        adapter.setOnItemSelectedListener(this);
+        layout.setAdapter(adapter);
+    }
+
+    private Item createItem(@StringRes int title, @DrawableRes int icon) {
+        var i = new Item();
+        i.setId(title);
+        i.setTitle(getText(title));
+        i.setIcon(getDrawable(icon));
+        return i;
+    }
+
+    // RecyclerItemAdapter.OnItemSelectedListener
+    @Override
+    public void onItemSelected(IItem iitem) {
+        Item item = (Item) iitem;
+        int id = item.getId();
+
+        if (id == R.string.duress_pwd_action_add || id == R.string.duress_pwd_action_update) {
+            var i = new Intent(this, DuressPasswordSetupActivity.class);
+            i.putExtra(DuressPasswordSetupActivity.EXTRA_TITLE_TEXT, id);
+            i.putExtra(DuressPasswordSetupActivity.EXTRA_USER_CREDENTIAL, userCredential);
+            allowNextOnStop = true;
+            startActivityForResult(i, REQ_CODE_SETUP);
+        } else if (id == R.string.duress_pwd_action_delete) {
+            var b = new AlertDialog.Builder(this);
+            b.setMessage(R.string.duress_pwd_action_delete_confirmation);
+            b.setPositiveButton(R.string.duress_pwd_delete_button, (dialog, which) -> {
+                LockPatternUtils lpu = getLockPatternUtils();
+                try {
+                    lpu.deleteDuressCredentials(userCredential);
+                } catch (Exception e) {
+                    Log.e(TAG, "deleteDuressCredentials failed", e);
+
+                    var d = new AlertDialog.Builder(this);
+                    d.setMessage(getString(R.string.duress_pwd_delete_error, e.toString()));
+                    d.setNeutralButton(R.string.duress_pwd_error_dialog_dismiss, null);
+                    d.show();
+                    return;
+                }
+                updateActionList();
+            });
+            b.setNegativeButton(R.string.duress_pwd_cancel_button, null);
+            b.show();
+        }
+    }
+
+    static final int REQ_CODE_OBTAIN_USER_CREDENTIALS = 1;
+    static final int REQ_CODE_SETUP = 2;
+
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        if (requestCode == REQ_CODE_SETUP) {
+            if (resultCode == RESULT_CANCELED) {
+                finish();
+            }
+            return;
+        }
+
+        if (requestCode != REQ_CODE_OBTAIN_USER_CREDENTIALS) {
+            throw new IllegalStateException(Integer.toString(resultCode));
+        }
+
+        Log.d(TAG, "onActivityResult, requestCode: " + requestCode + ", resultCode: " + resultCode);
+
+        if (resultCode != RESULT_OK) {
+            finish();
+            return;
+        }
+
+        if (data == null) {
+            throw new IllegalStateException("data == null");
+        }
+
+        var credential = data.getParcelableExtra(ChooseLockSettingsHelper.EXTRA_KEY_PASSWORD, LockscreenCredential.class);
+        if (credential == null) {
+            throw new IllegalStateException("no returned credential");
+        }
+        userCredential = credential;
+    }
+}
diff --git a/src/com/android/settings/security/DuressPasswordSetupActivity.java b/src/com/android/settings/security/DuressPasswordSetupActivity.java
new file mode 100644
index 0000000..9cedb11
--- /dev/null
+++ b/src/com/android/settings/security/DuressPasswordSetupActivity.java
@@ -0,0 +1,196 @@
+package com.android.settings.security;
+
+import android.app.AlertDialog;
+import android.os.Bundle;
+import android.util.Log;
+import android.view.KeyEvent;
+import android.view.inputmethod.EditorInfo;
+import android.widget.EditText;
+import android.widget.TextView;
+import android.widget.Toast;
+
+import androidx.annotation.IdRes;
+import androidx.annotation.Nullable;
+import androidx.annotation.StringRes;
+
+import com.android.internal.widget.LockPatternUtils;
+import com.android.internal.widget.LockscreenCredential;
+import com.android.internal.widget.PasswordValidationError;
+import com.android.settings.R;
+import com.android.settings.password.ChooseLockPassword.ChooseLockPasswordFragment.PasswordValidationErrorConverter;
+import com.google.android.setupcompat.template.FooterBarMixin;
+import com.google.android.setupcompat.template.FooterButton;
+import com.google.android.setupdesign.GlifLayout;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Function;
+
+public class DuressPasswordSetupActivity extends DuressPasswordActivity
+        implements TextView.OnEditorActionListener {
+    static final String TAG = DuressPasswordSetupActivity.class.getSimpleName();
+    static final String EXTRA_USER_CREDENTIAL = "user_credential";
+    static final String EXTRA_TITLE_TEXT = "title";
+    private boolean isUpdate;
+
+    enum DuressCredentialType {
+        PIN(LockscreenCredential::createPin,
+                R.id.pin_input, R.id.pin_input_confirmation,
+                R.string.lockpassword_confirm_pins_dont_match),
+        PASSWORD(LockscreenCredential::createPassword,
+                R.id.password_input, R.id.password_input_confirmation,
+                R.string.lockpassword_confirm_passwords_dont_match),
+        ;
+
+        final Function<String, LockscreenCredential> credenialCreator;
+        final @IdRes int mainInputId;
+        final @IdRes int confirmationInputId;
+        final @StringRes int confirmationMismatchText;
+
+        DuressCredentialType(Function<String, LockscreenCredential> credenialCreator,
+                             int mainInputId, int confirmationInputId,
+                             int confirmationMismatchText) {
+            this.credenialCreator = credenialCreator;
+            this.mainInputId = mainInputId;
+            this.confirmationInputId = confirmationInputId;
+            this.confirmationMismatchText = confirmationMismatchText;
+        }
+
+        LockscreenCredential createCredential(DuressPasswordSetupActivity activity) {
+            EditText ed = activity.requireViewById(mainInputId);
+            return credenialCreator.apply(ed.getText().toString());
+        }
+    }
+
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        setContentView(R.layout.duress_password_setup);
+
+        GlifLayout layout = requireViewById(R.id.glif_layout);
+        int headerText = getIntent().getExtras().getNumber(EXTRA_TITLE_TEXT);
+        isUpdate = headerText == R.string.duress_pwd_action_update;
+        layout.setHeaderText(headerText);
+        adjustDescriptionStyle(layout);
+
+        FooterBarMixin footerBar = layout.getMixin(FooterBarMixin.class);
+        {
+            var b = new FooterButton.Builder(this);
+            b.setText(isUpdate ? R.string.duress_pwd_update_button : R.string.duress_pwd_add_button);
+            b.setButtonType(FooterButton.ButtonType.DONE);
+            b.setListener(v -> save());
+            b.setTheme(com.google.android.setupdesign.R.style.SudGlifButton_Primary);
+            footerBar.setPrimaryButton(b.build());
+        }
+        {
+            var b = new FooterButton.Builder(this);
+            b.setText(R.string.duress_pwd_cancel_button);
+            b.setButtonType(FooterButton.ButtonType.CANCEL);
+            b.setListener(v -> finish());
+            b.setTheme(com.google.android.setupdesign.R.style.SudGlifButton_Secondary);
+            footerBar.setSecondaryButton(b.build());
+        }
+        for (var ct : DuressCredentialType.values()) {
+            for (int id : new int[] { ct.mainInputId, ct.confirmationInputId }) {
+                EditText ed = requireViewById(id);
+                ed.setTag(ct);
+                ed.setOnEditorActionListener(this);
+            }
+        }
+    }
+
+    // TextView.OnEditorActionListener
+    @Override
+    public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
+        if ((actionId != EditorInfo.IME_ACTION_NEXT && actionId != EditorInfo.IME_ACTION_DONE)) {
+            return false;
+        }
+
+        EditText ed = (EditText) v;
+        DuressCredentialType ct = (DuressCredentialType) ed.getTag();
+        int id = v.getId();
+
+        if (id == ct.mainInputId) {
+            return !checkCredentialInputErrors(ct);
+        }
+        if (id == ct.confirmationInputId) {
+            return !checkCredentialConfirmationError(ct);
+        }
+        return false;
+    }
+
+    private void save() {
+        boolean credentialCheckRes = true;
+        for (DuressCredentialType ct : DuressCredentialType.values()) {
+            credentialCheckRes &= checkCredentialInputErrors(ct);
+            credentialCheckRes &= checkCredentialConfirmationError(ct);
+        }
+        if (!credentialCheckRes) {
+            return;
+        }
+
+        var userCredential = getIntent().getParcelableExtra(EXTRA_USER_CREDENTIAL, LockscreenCredential.class);
+        Objects.requireNonNull(userCredential, EXTRA_USER_CREDENTIAL);
+        LockscreenCredential pin = DuressCredentialType.PIN.createCredential(this);
+        LockscreenCredential password = DuressCredentialType.PASSWORD.createCredential(this);
+        LockPatternUtils lockPatternUtils = getLockPatternUtils();
+
+        Runnable save = () -> {
+            try {
+                lockPatternUtils.setDuressCredentials(userCredential, pin, password);
+            } catch (Exception e) {
+                Log.e(TAG, "setDuressCredentials failed", e);
+
+                var d = new AlertDialog.Builder(this);
+                d.setMessage(getString(R.string.duress_pwd_save_error, e.toString()));
+                d.setNeutralButton(R.string.duress_pwd_error_dialog_dismiss, null);
+                d.show();
+                return;
+            }
+            Toast.makeText(this,
+                    isUpdate ? R.string.duress_pwd_toast_updated : R.string.duress_pwd_toast_added,
+                    Toast.LENGTH_LONG).show();
+            finish();
+        };
+
+        if (isUpdate) {
+            save.run();
+        } else {
+            var warning = new AlertDialog.Builder(this);
+            warning.setTitle(R.string.duress_pwd_save_warning_title);
+            warning.setMessage(R.string.duress_pwd_save_warning_text);
+            warning.setNegativeButton(R.string.duress_pwd_cancel_button, null);
+            warning.setPositiveButton(R.string.duress_pwd_proceed_button, (d, w) -> {
+                save.run();
+            });
+            warning.show();
+        }
+    }
+
+    private boolean checkCredentialInputErrors(DuressCredentialType ct) {
+        EditText ed = requireViewById(ct.mainInputId);
+        String text = ed.getText().toString();
+        LockscreenCredential c = ct.credenialCreator.apply(text);
+
+        List<PasswordValidationError> errors = LockPatternUtils.validateDuressCredential(c);
+        String error = null;
+        boolean res = true;
+        if (!errors.isEmpty()) {
+            res = false;
+            var pvec = new PasswordValidationErrorConverter(this, c.isPassword(), errors);
+            error = String.join("\n", pvec.convertErrorCodeToMessages());
+        }
+        ed.setError(error);
+        return res;
+    }
+
+    private boolean checkCredentialConfirmationError(DuressCredentialType ct) {
+        EditText main = requireViewById(ct.mainInputId);
+        EditText confirmation = requireViewById(ct.confirmationInputId);
+
+        boolean res = main.getText().toString().equals(confirmation.getText().toString());
+        confirmation.setError(res ? null : getText(ct.confirmationMismatchText));
+        return res;
+    }
+}
-- 
2.43.0

